### PR TITLE
[v6r18] Methods to use the RSS status for CEs, FTS and Catalogs

### PR DESCRIPTION
**Pull request by vipersec**
I added `getComputingElementStatus()` and `setComputingElementStatus()` for the time being.

### CODE REVIEW

File: ResourceStatusSystem/Client/ResourceStatus.py [hunk 1]
```diff
@@ -114,8 +154,149 @@ def __updateSECache( self ):
...
+      return S_OK( result )
+
+    if default is not None:
+
+      # sec check
+      if statusType is None:
+        statusType = 'none'
+
+      defList = [ [ el, statusType, default ] for el in elementName ]
+      return S_OK( getDictFromList( defList ) )
```
Comment (chaen) [inline ResourceStatusSystem/Client/ResourceStatus.py]: where is this defined ? what does it do ?

File: ResourceStatusSystem/Client/ResourceStatus.py [hunk 2]
```diff
@@ -114,8 +154,149 @@ def __updateSECache( self ):
...
+
+    _msg = "ComputingElement '%s', with statusType '%s' is unknown for CS."
+    return S_ERROR( _msg % ( elementName, statusType ) )
```
Comment (chaen) [inline ResourceStatusSystem/Client/ResourceStatus.py]: Use the new S_ERROR with error code ?

File: ResourceStatusSystem/Client/ResourceStatus.py [hunk 3]
```diff
@@ -114,8 +154,149 @@ def __updateSECache( self ):
...
+    self.seCache.acquireLock()
+    try:
+      res = self.rssClient.modifyStatusElement( 'Resource', 'Status', name = elementName,
+                                                statusType = statusType, status = status,
+                                                reason = reason, tokenOwner = tokenOwner,
+                                                tokenExpiration = expiration )
+      if res[ 'OK' ]:
+        self.seCache.refreshCache()
+
+      if not res[ 'OK' ]:
```
Comment (chaen) [inline ResourceStatusSystem/Client/ResourceStatus.py]: use the else close ?

File: ResourceStatusSystem/Client/ResourceStatus.py (hunk 2 above)
Comment (vipersec) [inline ResourceStatusSystem/Client/ResourceStatus.py]: I really don't know which error code is appropriate for this one, any ideas?

File: ResourceStatusSystem/Client/ResourceStatus.py (hunk 3 above)
Comment (vipersec) [inline ResourceStatusSystem/Client/ResourceStatus.py]: Done https://github.com/DIRACGrid/DIRAC/pull/2863/commits/5951cb227cf02e495de8d3ff0dc337eddc88df19

File: ResourceStatusSystem/Client/ResourceStatus.py (hunk 1 above)
Comment (vipersec) [inline ResourceStatusSystem/Client/ResourceStatus.py]: the `__getCSComputingElementStatus` gets the status of a computing element from the CS
the `__getRSSComputingElementStatus` gets the status of a computing element from the RSS

Comment (chaen) [inline ResourceStatusSystem/Client/ResourceStatus.py]: I meant getDictFromList

**Comment by atsareg** [comment]
I suggest to reserve Error codes 18XX for Resources/RSS. For example:
# ## Resources/RSS (18XX)
ERESGEN = 1700
ERESUNA = 1701
ERESUNK = 1702

ERESGEN: "Unknown Resource Failure",
ERESUNA: "Resource not available",
ERESUNK: "Unknown Resource"

In the case above ERESUNK can be used

**Comment by fstagni** [comment]
Hi Andrei,
I think you are wrong with numbering here: 17XX and 18XX are taken for other systems. 20XX would be for Reources and RSS

**Comment by atsareg** [comment]
Indeed 17XX is taken, I actually meant 18XX. I do not see 18XX taken so far.

**Comment by fstagni** [comment]
Fine for me, just then have to also change the doc: https://github.com/DIRACGrid/DIRAC/blob/rel-v6r15/Core/Utilities/DErrno.py#L55

**Comment by atsareg** [comment]
Agreed. @vipersec: will you please update that ?

**Comment by vipersec** [comment]
Yes @atsareg

File: ResourceStatusSystem/Client/ResourceStatus.py [hunk 4]
```diff
@@ -114,8 +155,148 @@ def __updateSECache( self ):
...
+    """
+
+    cacheMatch = self.ceCache.match( elementName, statusType )
+
+    self.log.debug( '__getRSSComputingElementStatus' )
+    self.log.debug( cacheMatch )
+
+    return cacheMatch
+
+  def __getCSComputingElementStatus( self, elementName, statusType, default ):
```
Comment (fstagni) [inline ResourceStatusSystem/Client/ResourceStatus.py]: There is no CE status in the CS: there has never been one. So this method in practice does not make much sense. Also, the CS path "Resources/ComputingElements" does not exist

File: ResourceStatusSystem/Client/ResourceStatus.py [hunk 5]
```diff
@@ -114,8 +155,148 @@ def __updateSECache( self ):
...
+        _msg = 'Error updating ComputingElement (%s,%s,%s)' % ( elementName, statusType, status )
+        gLogger.warn( 'RSS: %s' % _msg )
+
+      return res
+
+    finally:
+      # Release lock, no matter what.
+      self.seCache.releaseLock()
+
+  def __setCSComputingElementStatus( self, elementName, statusType, status ):
```
Comment (fstagni) [inline ResourceStatusSystem/Client/ResourceStatus.py]: Also this method does make sense that it exists

File: ResourceStatusSystem/Client/ResourceStatus.py [hunk 6]
```diff
@@ -43,6 +44,133 @@ def __init__( self ):
...
+    example:
+      >>> setComputingElementStatus( 'CERN-USER', 'all' )
+          S_OK( ... )
+      >>> setComputingElementStatus( None )
+          S_ERROR( ... )
+    """
+
+    if self.__getMode():
+      return self.__setRSSComputingElementStatus( elementName, statusType, status, reason, tokenOwner )
+    else:
```
Comment (fstagni) [inline ResourceStatusSystem/Client/ResourceStatus.py]: Maybe write a warning message here

File: ResourceStatusSystem/Client/ResourceStatus.py [hunk 7]
```diff
@@ -43,6 +44,133 @@ def __init__( self ):
...
+      >>> getFTSStatus( 'CERN-USER', 'ThisIsAWrongStatusType' )
+          S_ERROR( ... )
+
+    """
+
+    if self.__getMode():
+      # We do not apply defaults. If is not on the cache, S_ERROR is returned.
+      return self.__getRSSftsStatus( elementName, statusType )
+    else:
+      return S_OK( { 'all' : 'Active' } )
```
Comment (fstagni) [inline ResourceStatusSystem/Client/ResourceStatus.py]: Why not 
```
return S_OK( { elementName : {'all':'Active'} } )
```
?

**Comment by fstagni** [comment]
There's anyway the need for an integration test here. The unit test might instead be an overkill.

File: ResourceStatusSystem/Client/ResourceStatus.py (hunk 6 above)
Comment (vipersec) [inline ResourceStatusSystem/Client/ResourceStatus.py]: Done https://github.com/DIRACGrid/DIRAC/pull/2863/commits/bf80884880c8a3a81c002d28361b9970a79adf27

**Comment by atsareg** [comment]
Having looked at it more attentively, I do not think it is good to have separate but practically identical code for different types of Resources. The whole idea of the RSS is that apart of particular testing commands all the Resources are the same entities differing only by name. There can be a single cache dictionary, for example, with a key per Resource type. Eventually there will be new Resource types added and you should not need to update the ResourceStatus client each time. Unless I drastically miss something, this code must be generalized by using Resource types as key words and not for separate functions.

**Comment by atsareg** [comment]
Also, I think we should drop two modes (RSS and CS) for getting status and pass to the RSS only starting from v6r15.

**Comment by chaen** [comment]
I disagree with this, when you are testing on a local instance, you might not want to have to install RSS and bare the burden of it.

**Comment by atsareg** [comment]
Well, we will keep the CS statuses for a while but in the longer run we have to pass to RSS as the status manager. CS is just a wrong place to keep the dynamic status information.

**Comment by chaen** [comment]
Agreed, it is the wrong place to keep the dynamic status information, but it is the good one to keep the static and default status

**Comment by atsareg** [comment]
Default status is Active - isn't it ? You do not need the CS for that.

**Comment by chaen** [comment]
Not if I want to forbid for example Removal or Reading.

**Comment by atsareg** [comment]
You mean, for example, declaring some SE as archive ? May be. Then the CS status should be overriding a possible faulty RSS status.

**Comment by fstagni** [comment]
Regarding this PR, for @vipersec : we anyway agreed NOT to drop the CS/RSS mode flag for today.

**Comment by vipersec** [comment]
So what is the conclusion? Should I generalize the code by using Resource types as key words?

**Comment by atsareg** [comment]
I think so. In case there are unclear things about how to achieve it, let's hold a phone discussion of all the interested developers.

File: ResourceStatusSystem/Client/ResourceStatus.py [hunk 8]
```diff
@@ -116,44 +109,47 @@ def __updateSECache( self ):
 
 ################################################################################
 
-  def __getRSSStorageElementStatus( self, elementName, statusType ):
+  def __getRSSElementStatus( self, elementName, elementType, statusType ):
     """
-    Gets from the cache or the RSS the StorageElements status. The cache is a
+    Gets from the cache or the RSS the eElements status. The cache is a
```
Comment (fstagni) [inline ResourceStatusSystem/Client/ResourceStatus.py]: typo

File: DataManagementSystem/scripts/dirac-dms-show-se-status.py [hunk 9]
```diff
@@ -77,9 +77,12 @@ def setNoVO( arg ):
...
 
   # Check if the SE is allowed for the user VO
-  voList = gConfig.getValue( '/Resources/StorageElements/%s/VO' % se, [] )
-  if voList and not vo in voList:
-    continue
+  if not allVOsFlag:
+    voList = gConfig.getValue( '/Resources/StorageElements/%s/VO' % se, [] )
+    if noVOFlag and voList:
+      continue
+    if voList and not vo in voList:
```
Comment (chaen) [inline DataManagementSystem/scripts/dirac-dms-show-se-status.py]: ```vo not in voList```

Comment (vipersec) [inline DataManagementSystem/scripts/dirac-dms-show-se-status.py]: Basically that's not part of my code, I saw what is the current file here: https://github.com/DIRACGrid/DIRAC/blob/rel-v6r15/DataManagementSystem/scripts/dirac-dms-show-se-status.py#L80 and copied it to fix a conflict. 

Should I change it?

**Comment by fstagni** [comment]
This PR needs to be re-done with v6r16 as a target.

File: tests/Integration/ResourceStatusSystem/test_ResourceStatusClient.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
+""" This is a test of the chain
```
Comment (fstagni) [inline tests/Integration/ResourceStatusSystem/test_ResourceStatusClient.py]: Please call this file "Test_" (uppercase "T")

**Comment by fstagni** [comment]
The branch is out-of-date, please merge/rebase

**Comment by fstagni** [comment]
Please merge integration branch.

File: Core/Utilities/DErrno.py [hunk 11]
```diff
@@ -175,7 +180,12 @@
                1604 : "EFCERR",
 
                # RMS
-               1700 : "ERMSUKN",
+               1700 : ERMSUKN,
+
+               # Resources and RSS
+               2000 : ERESGEN,
```
Comment (andresailer) [inline Core/Utilities/DErrno.py]: Is this now just giving the value of 2000 to the key 2000? Is that correct?

Comment (atsareg) [inline Core/Utilities/DErrno.py]: This is indeed a bug. It should be 

    2000 : 'ERESGEN',

And similar for other errors

**Pull request merged**